crates/worker/src/partition_processor_manager/mod.rs

```diff
@@ -666,39 +734,82 @@ impl PartitionProcessorManager {
-                        .tx
-                        .try_send(ProcessorsManagerCommand::CreateSnapshot(*partition_id, tx))
-                        .is_err()
-                    {
-                        break;
+        let snapshot_partitions: Vec<_> = self
+            .processor_states
+            .iter()
+            .filter_map(|(partition_id, state)| {
+                state
+                    .partition_processor_status()
+                    .map(|status| (*partition_id, status))
+            })
+            .filter(|(_, status)| {
+                status.effective_mode == RunMode::Leader
+                    && status.replay_status == ReplayStatus::Active
+                    && status.last_applied_log_lsn.unwrap_or(Lsn::INVALID)
+                        >= status
+                            .last_archived_log_lsn
+                            .unwrap_or(Lsn::OLDEST)
+                            .add(Lsn::from(records_per_snapshot.get()))
+            })
+            .collect();
+
+        for (partition_id, status) in snapshot_partitions {
+            debug!(
+                %partition_id,
+                last_archived_lsn = %status.last_archived_log_lsn.unwrap_or(SequenceNumber::OLDEST),
+                last_applied_lsn = %status.last_applied_log_lsn.unwrap_or(SequenceNumber::INVALID),
+                "Requesting partition snapshot",
+            );
+            self.spawn_create_snapshot_task(partition_id, None);
+        }
+    }
+
+    /// Spawn a task to create a snapshot of the given partition. Optionally, a sender will be
+    /// notified of the result on completion.
+    fn spawn_create_snapshot_task(
+        &mut self,
+        partition_id: PartitionId,
+        sender: Option<oneshot::Sender<SnapshotResult>>,
+    ) {
+        if let Entry::Vacant(entry) = self.pending_snapshots.entry(partition_id) {
+            let config = self.updateable_config.live_load();
+
+            let snapshot_base_path = config.worker.snapshots.snapshots_dir(partition_id);
+            let snapshot_id = SnapshotId::new();
+
+            let create_snapshot_task = SnapshotPartitionTask {
+                snapshot_id,
+                partition_id,
+                snapshot_base_path,
+                partition_store_manager: self.partition_store_manager.clone(),
+                cluster_name: config.common.cluster_name().into(),
+                node_name: config.common.node_name().into(),
+            };
+
+            let spawn_task_result = restate_core::task_center().spawn_unmanaged(
+                TaskKind::PartitionSnapshotProducer,
+                "create-snapshot",
+                Some(partition_id),
+                create_snapshot_task.run(),
+            );
+
+            match spawn_task_result {
+                Ok(handle) => {
+                    self.snapshot_export_tasks.push(handle);
+                    if let Some(sender) = sender {
+                        entry.insert(sender);
                     }
```

Comment:
What if `sender == None`? Will this mean that we create a snapshot task but don't remember that it is in progress?

Reply:
I now explicitly call with `sender` = `None` when requesting a snapshot from the `PartitionProcessorManager`. We still have a handle to the task in `snapshot_export_tasks`, but there's no one outside of the PPM to notify about the result. A oneshot channel is now only used to respond to a `CreateSnapshot` RPC.

Reply:
Can't it then happen that we start multiple snapshot tasks for the same partition? I was under the impression that we wanted to allow only a single snapshot task per partition at any point in time.

Reply:
Doh, this was a silly bug to introduce at the last minute! Thanks for catching this.